Examples: Add Darcy flow example

sashaowen opened the pull request:
This PR adds a notebook showing an example of mapping permeability fields to pressure flow using the Darcy flow equation, which is a single-phase porous media flow equation.

mloubout commented on examples/cfd/09_Darcy_flow_equation.ipynb:
```diff
@@ -0,0 +1,436 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    class GaussianRF(object):
</pre><p>Any reason those use <code>torch</code> instead of <code>numpy</code>.I understand that your application may be ML related but this is a larger dependencies so I would stick to <code>numpy</code> for this tutorial that implements all these methods as well.   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

mloubout replied:
<p>Maybe show a few different samples to show what this generator does?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

mloubout replied:
<p>why not <code>-div(a * grad(u))</code> to match the equation? <code>a</code> is spatially dependent so it shouldn't factor out. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

mloubout commented:
Thanks @ sashaowen for the  verynice addition to the tutorials. I left a few comments but it shouldn't be too much to fix.

We usually run all the tutorials through the CI, so it would be good to
- FIx the random seed so that the result is always the same
- Add an assert on the norm of the result so that we know it is always running correctly
- add the `#NBVAL_IGNORE_OUTPUT` to cells generating figures or such to avoid CI errors.

let us know if you need some help with it.

sashaowen commented:
Hi, thanks for your replies, I have made the following changes:
- converted the GRF class to numpy and removed torch from imports
- Included three samples instead of one
- fixed the equation
- fixed a random seed 
- added an assert on the norm of the results
- added #NBVAL_IGNORE_OUTPUT to the cells generating figures 
Thanks!

georgebisbas commented on examples/cfd/09_Darcy_flow_equation.ipynb:
```diff
@@ -0,0 +1,538 @@
+{
```
<p>Comments should start with capital</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

georgebisbas replied:
<p>Boundary conditions</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

georgebisbas replied:
<p>using Devito</p><p>(add empty line)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

sashaowen commented:
Thanks @georgebisbas, the latest commit should have the changes mentioned in your comments :-)

mloubout commented on examples/cfd/09_Darcy_flow_equation.ipynb:
```diff
@@ -0,0 +1,537 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #17.        a.data[:] = perm[:]
</pre><p><code>initialize_function</code> is recommended here for <code>a</code> at least so that the FD halo is padded with the edge values because currently it assumes that <code>a</code> is zero (not physical and potentially numerically dangerous).</p><p> </p><p><code>initialize_function(a, perm, 0)</code> should do it</p><p> </p><p>typo: pseudo not psuedo</p><p> </p><p>"the same data"? Do you mean initialize <code>a, f</code> with inputs permeability and forcing? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

mloubout replied:
<p>Remove print statement to avoid CI issue if a decimal doesn't match</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1998/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

mloubout commented:
Left some extra minor comments but is almost there, thanks for the patience.

Please rebase on top of master and try to follow commit message guidelines:
https://github.com/devitocodes/devito/wiki/Tags-for-commit-messages-and-PR-titles

sashaowen commented:
> Left some extra minor comments but is almost there, thanks for the patience.
> 
> Please rebase on top of master and try to follow commit message guidelines: https://github.com/devitocodes/devito/wiki/Tags-for-commit-messages-and-PR-titles

Thanks for your comments!,
Hopefully I have followed the commit guidelines properly this time, apologies for the previous ones.
 I have made the changes mentioned in your comments, however I'm unsure what rebase on top of master means, could you please point me to a resource on how to do it? Thanks!

mloubout commented:
Sorry for the delay. `rebase` lets you rewrite your commit history and make sure that your branch is up to date with master and is mergeable. 

You can simply run `git rebase -i origin/master` (ou'll need your fork to be up to date on master) and it will pop up an editable message with guidelines on what to do.

sashaowen commented:
Hi @mloubout I think I have addressed your comments, would you mind taking another look when you get the chance please? Hopefully I have done it right!

mloubout commented:
Hi. It looks good now. Your branch is out of date with master and need to be rebased and will be good to go
https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase#:~:text=What%20is%20git%20rebase%3F,of%20a%20feature%20branching%20workflow.

sashaowen commented:
> Hi. It looks good now. Your branch is out of date with master and need to be rebased and will be good to go https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase#:~:text=What%20is%20git%20rebase%3F,of%20a%20feature%20branching%20workflow.

I think this is up to date now? + rebased, I think, thankyou for your patience!!

mloubout commented:
Thanks, all good indeed. Looks liek the assert is failing in the notebook though

```
----> 1 assert np.isclose(LA.norm(output1),1.0335084)
      2 assert np.isclose(LA.norm(output2),1.3038709)
      3 assert np.isclose(LA.norm(output3),1.3940924)
```

Maybe a tolerance would make it more stable , for example `np.isclose(a, b, atol=1e-4, rtol=0)` should be ok usually

sashaowen commented:
@mloubout Added a tolerance and seemed to run fine locally so hopefully that should work now :)

sashaowen commented:
I just ran the same code separately in a jupyter server compared to VS code and got slightly different numbers, i.e

output 1 in VS code: 1.0335084

output 1 in a Jupyter server: 1.0335636

running the assert in the Jupyter server failed for 1e-4 but worked for 1e-3, I could try changing to 1e-3 if that is acceptable?

mloubout commented:
Yes

sashaowen commented:
committed with 1e-3.
